### PR TITLE
Add tooltip to indicator

### LINKED ISSUE #137: Add tooltip to indicator icon
## Problem

A tooltip could provide some extra information before clicking or just some quicker information that feels faster to access without clicking at all.

## Proposal

Add a tooltip with the status it is indicating, and maybe the middle-click action:

>**Bluetooth: On, connected to “Device Name”**
>Middle-click to disconnect and disable Bluetooth

## Prior Art

We added a tooltip to Applications Menu with a longer description and keyboard shortcuts.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/94658730-add-tooltip-to-indicator-icon?utm_campaign=plugin&utm_content=tracker%2F59641246&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F59641246&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by andirsun**
Fixes #137 

~THIS PR MUST TO BE MERGED FIRST : https://github.com/elementary/granite/pull/446~

### Demo :
Screenshot from 2020-12-02 11-50-48

### Tasks : 
- [x] Get Bluetooth state
- [x] Add Secondary tooltip message
- [x] Get Bluetooth device name

### CODE REVIEW

File: src/Indicator.vala
Comment (meisenzahl) [inline src/Indicator.vala:79]:
```suggestion
        string bluetooth_state = _("Off");
```

File: src/Indicator.vala
Comment (meisenzahl) [inline src/Indicator.vala:82]:
```suggestion
            bluetooth_state = _("On");
```

**Review by meisenzahl — CHANGES_REQUESTED** [review]

**Review by cassidyjames — CHANGES_REQUESTED** [review]
Looking at this in context, I think we can simplify what's shown.

## On, no devices:

> Bluetooth Enabled

## On, connected to device

>**Bluetooth connected to "Device Name"**
>Middle-click to disable

## Off

>**Bluetooth Disabled**
>Middle-click to enable

---

Alternately, the small text could always read "Middle-click to toggle"

**Comment by andirsun** [comment]
@cassidyjames Hi!, I dont know how to pango markup works,
I am noticed that 
```
<Alt>Home //works 
```
but
```
<small> Some example test </small>
```
doesnt show any text, what Im missing ?

**Comment by andirsun** [comment]
Thats what looks like
Screenshot from 2020-12-08 08-58-58

File: src/Indicator.vala
Comment (cassidyjames) [inline src/Indicator.vala:79]: I think it's best (and easier for translators) to be less clever here, and just use the entire string.

```suggestion
        string description = _("Bluetooth is off");
```

File: src/Indicator.vala
Comment (cassidyjames) [inline src/Indicator.vala:83]:
```suggestion
            description = _("Bluetooth connected to %s".printf (paired_device.name));
```

File: src/Indicator.vala
Comment (cassidyjames) [inline src/Indicator.vala:85]:
```suggestion
            description = _("Bluetooth is on");
```

File: src/Indicator.vala
Comment (cassidyjames) [inline src/Indicator.vala:93]:
```suggestion
            description, Granite.TOOLTIP_SECONDARY_TEXT_MARKUP.printf (context)
```

**Review by cassidyjames — DISMISSED** [review]
This looks close to me. Just a suggestion for being a little less clever with the Bluetooth string.

**Review by cassidyjames — COMMENTED** [review]
I'm happy with this UX-wise but would like a code review from @elementary/desktop-developers

**Comment by danrabbit** [comment]
I'm not sure it really makes sense to call out being connected to a specific device. If I'm connected to speakers and a keyboard and a mouse and my phone, then what? Maybe it would be better to just say "Bluetooth is in use" or "Bluetooth devices are connected"

**Comment by cassidyjames** [comment]
@danrabbit hm that's fair. Coming from my phone, the Bluetooth icon shows when an audio device is paired by name—that's particularly handy to know you're paired to your buds versus a Bluetooth speaker, for example. I'm not sure what it does for other input devices, or how we would best handle multiple devices.

**Comment by cassidyjames** [comment]
Okay, yeah I tested this branch with a few devices. When you just have audio devices, it's super nice to be able to see which one it's connected to. But then as soon as you mix another device into the mix, it looks like it just shows the most-recently-connected device in the tooltip, which is less useful.

Active audio output device is probably more useful for the Sound indicator if anywhere, so I would be fine to simplify this to a tri-state:

- on (but not connected to anything)
- off
- connected

**Comment by danrabbit** [comment]
Yeah I was just gonna say that just because a device is paired doesn't mean it's the active audio device and the sound indicator would know that info 👍

**Comment by andirsun** [comment]
> Okay, yeah I tested this branch with a few devices. When you just have audio devices, it's super nice to be able to see which one it's connected to. But then as soon as you mix another device into the mix, it looks like it just shows the most-recently-connected device in the tooltip, which is less useful.
> 
> Active audio output device is probably more useful for the Sound indicator if anywhere, so I would be fine to simplify this to a tri-state:
> 
> - on (but not connected to anything)
> - off
> - connected
> 
> …which, incidentally, is what the icon itself communicates as well.

Sure !, Let me work on this

File: src/Indicator.vala
Comment (pongloongyeat) [inline src/Indicator.vala:80]: I think it'll be better to have "Middle-click to disable Bluetooth" here instead of just "to disable". I personally think it'd be better to give a context, much like the other indicators where we now have "Middle-click to enable/disable DnD" (notifications) or "Middle-click to snooze" (nightlight).

Note that this does come with the setback of not looking super nice in languages that do not have an official wording for "Bluetooth". For instance, "disable Bluetooth" translates to "禁用 Bluetooth" which will have a mix of English and Mandarin so I'd suggest to wait for a response from the design team.

File: src/Indicator.vala
Comment (pongloongyeat) [inline src/Indicator.vala:88]: Same as above.

**Review by pongloongyeat — COMMENTED** [review]
Haven't actually tested but just some thoughts

File: src/Indicator.vala
Comment (pongloongyeat) [reply src/Indicator.vala:80]: I meant just line 80. Sorry, never done a review before.

File: src/Indicator.vala
Comment (pongloongyeat) [reply src/Indicator.vala:88]: Just line 88

File: src/Indicator.vala
Comment (pongloongyeat) [inline src/Indicator.vala:33]: You'll need to call update_tooltip () here to initialise your tooltip. Otherwise, the user will only see a tooltip when they pair a device.

```suggestion
        var settings = new Settings ("io.elementary.desktop.wingpanel.bluetooth");
        display_widget = new Widgets.DisplayWidget (object_manager);
        
        var state = settings.get_boolean ("bluetooth-enabled");
        update_tooltip (state, false);
```

**Review by pongloongyeat — CHANGES_REQUESTED** [review]

File: src/Indicator.vala
Comment (danrabbit) [inline src/Indicator.vala:45]: Looks like this isn't used anymore :)
```suggestion
```

File: src/Indicator.vala
Comment (danrabbit) [inline src/Indicator.vala:23]: That's fine to be more specific about namespace, but we should still explicitly define scope:
```suggestion
    private BluetoothIndicator.Widgets.PopoverWidget popover_widget;
```

File: src/Indicator.vala
Comment (danrabbit) [inline src/Indicator.vala:22]: Doesn't look like either of these are used anymore
```suggestion
```

File: src/Indicator.vala
Comment (danrabbit) [inline src/Indicator.vala:41]: Why was this moved to the constructor instead of staying in `get_display_widget ()`?

**Review by danrabbit — CHANGES_REQUESTED** [review]
Just a few comments/questions :)

File: src/Indicator.vala
Comment (danrabbit) [inline src/Indicator.vala:22]: explicit scope on this one too please :)
```suggestion
    private Widgets.DisplayWidget? display_widget;
```

**Comment by jeremypw** [comment]
@andirsun Are you still working on this or do you feel all the requested changes have been made/resolved (in which case I would dismiss the review with a reason as well as requesting a new one)?

**Comment by andirsun** [comment]
> @andirsun Are you still working on this or do you feel all the requested changes have been made/resolved (in which case I would dismiss the review with a reason as well as requesting a new one)?

Sure!, Im a little bussy for now, you can copy my branch in to other and finish de PR

File: src/Indicator.vala
Comment (pongloongyeat) [reply src/Indicator.vala:39]: I think this was my requested change though I'm not sure I remember why but this was what I wrote

> You'll need to call update_tooltip () here to initialise your tooltip. Otherwise, the user will only see a tooltip when they pair a device.

I'll check back on this in a bit.

Comment (pongloongyeat) [reply src/Indicator.vala:39]: Yeah I'm sorry, this was my bad. I'm not sure why I requested for it to be moved out of the constructor. I think what I wanted was just for `update_tooltip ()` to be called after the DisplayWidget is constructed but I'm not sure how my confusion led me to this. Sorry about this, I'll push a commit to revert this.

Comment (pongloongyeat) [reply src/Indicator.vala:39]: No wait, I didn't request for it to be moved out, just that `update_tooltip ()` be called after construction of the `DisplayWidget` so that the tooltip shows up instead of only showing after a device is paired.

Please ignore my comments in this conversation.

**Review by pongloongyeat — APPROVED** [review]

**Pull request closed without merging**
